feat: allow scaling of added elements

danielolsen opened the pull request:
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
Allow scaling to be applied to elements that are added (i.e., not in the base grid). Closes #464.

### What the code is doing
We separate the scaling functions into scaling-by-zone and scaling-by-ID. We apply scaling-by-zone, then additions, then scaling-by-ID.

### Testing
Tested on Scenario 3856, which was created using test code along these lines, so the change table that's on the server is not valid with the current implementation of TransformGrid. With the new code, it works.
```python
from powersimdata import Scenario
from powersimdata.design.transmission.mwmiles import calculate_mw_miles
scenario = Scenario(3856)
calculate_mw_miles(scenario)
```

### Time estimate
15 minutes.

BainanXia reviewed (approved):
This is consistent with our previous discussion.

danielolsen commented:
I added a test of the TransformGrid functionality, which fails initially but passes with the new functionality.

Importantly, for this test, we have to edit the change table directly, since the change table method won't let us manually scale a newly-created branch yet. So this code means that Scenario 3856 works, since the change table was already created via the scale congested mesh branches method, but that the same scaling could not be applied manually by the user.

BainanXia commented:
> I added a test of the TransformGrid functionality, which fails initially but passes with the new functionality.
> 
> Importantly, for this test, we have to edit the change table directly, since the change table method won't let us manually scale a newly-created branch yet. So this code means that Scenario 3856 works, since the change table was already created via the scale congested mesh branches method, but that the same scaling could not be applied manually by the user.

I think we should implement `self._get_new_ac_line()` and `self._get_new_dc_line()` follow the same logic of `self._get_new_bus()` in change table.

danielolsen commented:
> > I added a test of the TransformGrid functionality, which fails initially but passes with the new functionality.
> > Importantly, for this test, we have to edit the change table directly, since the change table method won't let us manually scale a newly-created branch yet. So this code means that Scenario 3856 works, since the change table was already created via the scale congested mesh branches method, but that the same scaling could not be applied manually by the user.
> 
> I think we should implement `self._get_new_ac_line()` and `self._get_new_dc_line()` follow the same logic of `self._get_new_bus()` in change table.

I got started on this, with `self._get_new_branch`. I used `branch` rather than `ac_line` since what we're fetching is a modified (if necessary) `branch` attribute data frame. I'll add `self._get_new_dcline` and `self._get_new_plant`, and I think that should complete the functionality.

danielolsen commented:
Caches for new {buses, branches, DC lines, and plants} are added.

BainanXia commented on powersimdata/input/change_table.py:
```diff
@@ -782,16 +781,19 @@ def add_bus(self, info):
-        new_bus_tuple = tuple(tuple(sorted(b.items())) for b in self.ct["new_bus"])
-        if new_bus_tuple in self.new_bus_cache:
-            return self.new_bus_cache[new_bus_tuple]
+    def _get_df_with_new_elements(self, table):
```
Missing doc strings for table and return values.

danielolsen replied:
Done.

BainanXia commented:
I've tested this with several corner cases that are made up myself. It works as expected. One minor concern is the cache can't be overwritten with new updates if one would like to track the changes step by step when building a change table, but that's not a typical user story.

danielolsen commented:
> I've tested this with several corner cases that are made up myself. It works as expected. One minor concern is the cache can't be overwritten with new updates if one would like to track the changes step by step when building a change table, but that's not a typical user story.

The original idea of the cache was that we wanted to make sure that we could add new elements step by step, without needing to instantiate a new TransformGrid each time (to get indices, voltages, interconnects etc. for the new buses), since the new bus info won't change as long as the `"new_bus"` entries in the change table don't change.

I'm not sure if I fully understand your user story, but I think for user who wants to track the grid changes step-by-step, they could call `scenario.get_grid()` between each change table method call. This cache and the function that uses it are not meant to be user-facing, which is why I added a leading underscore when I renamed the cache.

BainanXia commented:
> > I've tested this with several corner cases that are made up myself. It works as expected. One minor concern is the cache can't be overwritten with new updates if one would like to track the changes step by step when building a change table, but that's not a typical user story.
> 
> The original idea of the cache was that we wanted to make sure that we could add new elements step by step, without needing to instantiate a new TransformGrid each time (to get indices, voltages, interconnects etc. for the new buses), since the new bus info won't change as long as the `"new_bus"` entries in the change table don't change.
> 
> I'm not sure if I fully understand your user story, but I think for user who wants to track the grid changes step-by-step, they could call `scenario.get_grid()` between each change table method call. This cache and the function that uses it are not meant to be user-facing, which is why I added a leading underscore when I renamed the cache.

That's exactly what I'm referring to. I was lazy and didn't use `get_grid()` but using `_get_df_with_new_elements` instead and found the changes were not as expected. Then I realized that I shouldn't track changes in that way. That's what I mean by "not a typical user story". I think this is good to go.
